### PR TITLE
Gossiper: replace seastar threads with coroutines

### PR DESCRIPTION
**Pull request by bhalevy**
Many of the gossiper internal functions currently use seastar threads for historical reasons,
but since they are short living, the cost of spawning a seastar thread for them is excessive
and they can be simplified and made more efficient using coroutines.

### CODE REVIEW

**Comment by bhalevy** [comment]
I'd like to merge this PR before making further changes in those areas for https://github.com/scylladb/scylladb/issues/12282

File: gms/gossiper.cc [hunk 1]
```diff
@@ -2172,32 +2173,31 @@ future<> gossiper::add_local_application_state(std::initializer_list<std::pair<a
...
 future<> gossiper::add_local_application_state(std::list<std::pair<application_state, versioned_value>> states) {
     if (states.empty()) {
-        return make_ready_future<>();
+        co_return;
     }
-    return container().invoke_on(0, [states = std::move(states)] (gossiper& gossiper) mutable {
-        return seastar::async([g = gossiper.shared_from_this(), states = std::move(states)]() mutable {
-            auto& gossiper = *g;
+    try {
+        co_await container().invoke_on(0, [&] (gossiper& gossiper) mutable -> future<> {
```
Comment (avikivity) [inline gms/gossiper.cc:2186]: coroutine::lambda()?

File: gms/gossiper.cc [hunk 2]
```diff
@@ -214,25 +217,27 @@ future<> gossiper::handle_syn_msg(msg_addr from, gossip_digest_syn syn_msg) {
...
                         // No more pending syn msg to process
                         p.pending = false;
                         logger.debug("No more queued gossip syn msg from node {}, syn_msg={}, pending={}",
                                 from, p.syn_msg, p.pending);
-                        return stop_iteration::yes;
+                        co_return;
                     }
-                }).handle_exception([this, from] (std::exception_ptr ep) {
+        } catch (...) {
+            ep = std::current_exception();
```
Comment (avikivity) [inline gms/gossiper.cc:230]: Can't you declare ep here?

File: gms/gossiper.cc [hunk 3]
```diff
@@ -214,25 +217,27 @@ future<> gossiper::handle_syn_msg(msg_addr from, gossip_digest_syn syn_msg) {
...
                     if (_syn_handlers.contains(from.addr)) {
                         syn_msg_pending& p = _syn_handlers[from.addr];
                         p.pending = false;
                         p.syn_msg = {};
                     }
                     logger.warn("Failed to process gossip syn msg from node {}:  {}", from, ep);
-                    return make_exception_future<stop_iteration>(ep);
```
Comment (avikivity) [inline gms/gossiper.cc:231]: And return it here?

File: gms/gossiper.cc [hunk 4]
```diff
@@ -337,27 +341,28 @@ future<> gossiper::handle_ack_msg(msg_addr id, gossip_digest_ack ack_msg) {
...
-                    }).handle_exception([this, from] (std::exception_ptr ep) {
+        } catch (...) {
+            ep = std::current_exception();
                         if (_ack_handlers.contains(from.addr)) {
                             ack_msg_pending& p = _ack_handlers[from.addr];
                             p.pending = false;
                             p.ack_msg_digest = {};
                             logger.warn("Failed to process gossip ack msg digests from node {}: {}", from, ep);
                         }
-                        return make_exception_future<stop_iteration>(ep);
```
Comment (avikivity) [inline gms/gossiper.cc:349]: Similarly, here. Can't ep live in the catch clause? Makes everything simpler to read later.

File: gms/gossiper.cc [hunk 5]
```diff
@@ -1305,10 +1305,10 @@ future<> gossiper::unsafe_assassinate_endpoint(sstring address) {
 }
 
 future<> gossiper::assassinate_endpoint(sstring address) {
-    return container().invoke_on(0, [address] (auto&& gossiper) {
-        return seastar::async([&gossiper, g = gossiper.shared_from_this(), address] {
+    co_await container().invoke_on(0, [&] (auto&& gossiper) -> future<> {
```
Comment (avikivity) [inline gms/gossiper.cc:1308]: coroutine::lambda?

**Comment by avikivity** [comment]
Looks good, some minor comments.

File: gms/gossiper.cc (hunk 1 above)
Comment (bhalevy) [reply gms/gossiper.cc:2186]: I understand we need coroutine::lambda for continuations.
Is it needed for invoke_on / submit_to calls as well?

Is it needed when the lambda has no captures?

File: gms/gossiper.cc (hunk 3 above)
Comment (bhalevy) [reply gms/gossiper.cc:231]: I'll look into it.

**Comment by bhalevy** [comment]
v2 ([47dc287](https://github.com/scylladb/scylladb/commit/47dc287efd4f33a63b91fd9d77959f846bf3c3ed)):
- cleaned up includes that were added by the editor
- define `auto ep = std::current_exception()` in catch clauses, then throw
- added: `gossiper: handle_ack_msg: always log warning on exception`

File: gms/gossiper.cc (hunk 1 above)
Comment (avikivity) [reply gms/gossiper.cc:2174]: It's capturing &, so one variable.

I looked at submit_to() and it moves the lambda to a safe place, so it's not in the same category as a continuation. We should mark it as safe for continuations.

**Review by avikivity — APPROVED** [review]

**Pull request merged**